docs/source/validators.rst

```diff
@@ -0,0 +1,225 @@
+==========
+Validation
+==========
+
+.. module:: adapters.validation
+
+
+Validation Nodes
+================
+
+The main components of validation are instances of :class:`ValidationNode`. Each
+node contains a bit of validation logic that operates on a predefined set of
+data. Multiple nodes can be combined to a
+:class:`~adapters.validation.ValidationTree` to build more complex validation
+scenarios.
+
+Example: User Validation
+------------------------
+
+You can create validation nodes by subclassing
+:class:`~DeclarativeValidationNode`::
+
+    class ValidateUsername(DeclarativeValidationNode):
+        inputs = {'username'}
+
+We start with specifying which fields the node expects as input in the
+:attr:`~DeclarativeValidationNode.inputs` attribute. Here, the
+``ValidateUsername`` node expects only an username. This is everything that is
+needed to create a validation node. Of course, a validation node that doesn't
+actually validate anything isn't particularly useful. So let's now define
+validation functions::
+
+    class ValidateUsername(DeclarativeValidationNode):
+        inputs = {'username'}
+
+        @staticmethod
+        def validate_username_length(data, **kwargs):
+            if len(data['username']) < 5:
+                raise ValidationError(
+                    'username must have at least 5 characters',
+                    'username'
+                )
+
+        @staticmethod
+        def validate_username_starts_with_capital_letter(data, **kwargs):
+            if not data['username'][0].isupper():
```

Comment:
There's a minor bug here - since this doesn't depend on `validate_username_length`, it can be passed the empty string and would raise `IndexError` here.